src/main/java/com/puppycrawl/tools/checkstyle/utils/ChainedPropertyUtil.java

```diff
@@ -0,0 +1,133 @@
+
+/**
+ * Resolves chained properties from a user-defined property file.
+ */
+public final class ChainedPropertyUtil {
+
+    /**
+     * Used to report undefined property in exception message.
+     */
+    public static final String UNDEFINED_PROPERTY_MESSAGE = "Undefined property: ";
+
+    /**
+     * Property variable expression pattern, matches property variables such as {@code ${basedir}}.
+     */
+    private static final Pattern PROPERTY_VARIABLE_PATTERN = Pattern.compile("\\$\\{([^\\s}]+)}");
+
+    /**
+     * Prevent instantiation.
+     */
+    private ChainedPropertyUtil() {
+    }
+
+    /**
+     * Accepts user defined properties and returns new properties
+     * with all chained properties resolved.
+     *
+     * @param properties the underlying properties to use
+     *                   for property resolution.
+     * @return resolved properties
+     * @throws CheckstyleException when chained property is not defined
+     */
+    public static Properties getResolvedProperties(Properties properties)
+            throws CheckstyleException {
+        final Set<String> unresolvedPropertyNames =
+            new HashSet<>(properties.stringPropertyNames());
+        Iterator<String> unresolvedPropertyIterator = unresolvedPropertyNames.iterator();
+        final Map<Object, Object> comparisonProperties = new Properties();
+
+        while (unresolvedPropertyIterator.hasNext()) {
+            final String propertyName = unresolvedPropertyIterator.next();
+            String propertyValue = properties.getProperty(propertyName);
+            final Matcher matcher = PROPERTY_VARIABLE_PATTERN.matcher(propertyValue);
+
+            while (matcher.find()) {
+                final String propertyVariableExpression = matcher.group();
+                final String unresolvedPropertyName =
+                    getPropertyNameFromExpression(propertyVariableExpression);
+
+                final String resolvedPropertyValue =
+                    properties.getProperty(unresolvedPropertyName);
+
+                if (resolvedPropertyValue != null) {
+                    propertyValue = propertyValue.replace(propertyVariableExpression,
+                        resolvedPropertyValue);
+                    properties.setProperty(propertyName, propertyValue);
+                }
+            }
+
+            if (allChainedPropertiesAreResolved(propertyValue)) {
+                unresolvedPropertyIterator.remove();
+            }
+
+            if (!unresolvedPropertyIterator.hasNext()) {
+
+                if (comparisonProperties.equals(properties)) {
+                    // At this point, we will have not resolved any properties in two iterations,
+                    // so unresolvable properties exist.
+                    throw new CheckstyleException(UNDEFINED_PROPERTY_MESSAGE
+                        + unresolvedPropertyNames);
+                }
+                comparisonProperties.putAll(properties);
```

Comment:
After each complete iteration through`unresolvedPropertyNames`, `comparisonProperties` is updated.  Since `properties` and `comparisonProperties` are set to be equal at this point, if we go through another complete cycle, and they are still equal, we have resolved no further properties, indicating that unresolvable properties exist.